src/Stack/Config.hs

```diff
@@ -587,6 +604,14 @@ loadBuildConfig mproject maresolver mcompiler = do
             { projectResolver = fromMaybe (projectResolver project') mresolver
             }
 
+    lockFileOutdated <- isLockFileOutdated stackYamlFP
+    unless lockFileOutdated (logDebug "Lock file is upto date")
+    when lockFileOutdated (logDebug "Lock file is outdated" >> generateLockFile stackYamlFP)
+
+    -- liftIO $ resolveLockFile (parent stackYamlFP)
+    lockFile <- liftIO $ addFileExtension "lock" stackYamlFP
+    cachePL <- liftIO $ loadLockFile lockFile
```

Comment:
Down below in that linked thread, this is the discussion which happened:

>  qrilka: Should we exclude that requirement of a successful build then @snoyberg ? And use lock file only to lock package versions without giving extra guarantees?
> Michael: Yes, I think so, good catch.

So, my idea is here this:
* Check if lock file is outdated (It may have just become outdated because of change in `stack.yaml` file.)
* If it's outdated, then generate a new lock file.
* Once the lock file is generated, use that information in `loadBuildConfig`.

Reply:
Oh, it was actually me, that's funny :)
I think your solution could work but it looks a bit backwards to me: you write a lock file and then read it back and use for completing locations. And as a lock file and `stack.yaml` times could be out of sync you could get into a situation when you'll use a stale lock, complete locations but don't write the new ones into a new lock file.
Imo it's more logical to do all of that in 1 pass:
1. read stack.yaml and its lock file
2. check stack.yaml deps whether they need to be completed or just could be taken from the lock file.
3. if a list of completed entries has changed - write a new lock file
4. continue further with completed package locations
A side note: for `stack.yaml` it's not critical but could easily get important for large snapshots - if you compare original raw locations - why not to use a `Map` or a `HashMap` for faster lookups?

Reply:
@qrilka Assuming that such a stale lock scenario arises, the code will still work in those cases. In the function `stackCompletePackageLocation`, the first argument to it is a list whose content  is generated from the lock file. If we are not able to find the `RawPackageLocation` in the lock file, we use `completePackageLocation` to actually produce the right one. So, this code will work even in the presence of stale lock. Do you agree with it or is there any edge case I have missed ?

Reg `Map` or `HashMap`, I totally agree. I want to get other things right first! But please leave those comments also as I will get back to them once I got the fundamentals right.

Reply:
See above "your solution could work" :)
The end result should be the same and what concerns me more is that this lock-related logic is a bit split between different parts of the code, at least I'd propose at least to extract some function like `getCompletedLocations` with lock file related code in it and add a short text comment describing it maybe.

Reply:
I have changed the naming of the functions and also changed the cached data structure to Map. Hopefully that helps. Resolving this comment, free feel to unresolve this or any other of my resolved comments. :-)